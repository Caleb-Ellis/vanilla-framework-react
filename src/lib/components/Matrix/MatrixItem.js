import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

const MatrixItem = (props) => {
  const {
    children, className, href, img, title,
  } = props;

  const classNames = getClassName({
    [className]: className,
    'p-matrix__item': true,
  }) || undefined;

  return (
    <li className={classNames}>
      {img && <img className="p-matrix__img" src={img.src} alt={img.alt} />}
      <div className="p-matrix__content">
        {title &&
          <h3 className="p-matrix__title">
            {href ? <a className="p-matrix__link" href={href}>{title}</a> : title}
          </h3>}
        <div className="p-matrix__desc">
          { children }
        </div>
      </div>
    </li>
  );
};

MatrixItem.defaultProps = {
  children: null,
  className: undefined,
  href: '',
  img: null,
  title: '',
};

MatrixItem.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  img: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string,
  }),
  href: PropTypes.string,
  title: PropTypes.string,
};

MatrixItem.displayName = 'MatrixItem';

export default MatrixItem;
