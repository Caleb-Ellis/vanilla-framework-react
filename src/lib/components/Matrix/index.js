export { default as Matrix } from './Matrix';
export { default as MatrixItem } from './MatrixItem';
