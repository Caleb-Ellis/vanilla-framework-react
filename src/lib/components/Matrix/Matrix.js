import React from 'react';

import MatrixItem from './MatrixItem';

class Matrix extends React.Component {
  constructor() {
    super();
    this.generateExtras = this.generateExtras.bind(this);
  }

  generateExtras() {
    const { children } = this.props;
    const emptyItemCount = (2 * React.Children.count(children)) % 3;
    const extraMatrixItems = [];

    for (let i = 0; i < emptyItemCount; i += 1) {
      extraMatrixItems.push(<MatrixItem key={i} className="u-hide--small" />);
    }

    return extraMatrixItems;
  }

  render() {
    const { children } = this.props;

    // Generate extra MatrixItems to fill empty space in a row
    const extraMatrixItems = this.generateExtras();

    return (
      <ul className="p-matrix u-clearfix">
        { children }
        { extraMatrixItems.length > 0 && extraMatrixItems }
      </ul>
    );
  }
}

Matrix.defaultProps = {
  children: null,
};

Matrix.propTypes = {
  children: (props, propName, componentName) => {
    const prop = props[propName];
    let error = null;

    React.Children.forEach(prop, (child) => {
      if (child.type !== MatrixItem) {
        error = new Error(`${componentName} children should be of type "MatrixItem".`);
      }
    });

    return error;
  },
};

Matrix.displayName = 'Matrix';

export default Matrix;
