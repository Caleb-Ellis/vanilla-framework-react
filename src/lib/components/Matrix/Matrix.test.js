import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Matrix from './Matrix';
import MatrixItem from './MatrixItem';

describe('Matrix component', () => {
  it('should render a Matrix with one MatrixItem correctly', () => {
    const matrix = ReactTestRenderer.create(
      <Matrix>
        <MatrixItem>
          description
        </MatrixItem>
      </Matrix>,
    );
    const json = matrix.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a Matrix with multiple MatrixItem correctly', () => {
    const matrix = ReactTestRenderer.create(
      <Matrix>
        <MatrixItem>
          description
        </MatrixItem>
        <MatrixItem>
          description
        </MatrixItem>
        <MatrixItem>
          description
        </MatrixItem>
      </Matrix>,
    );
    const json = matrix.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a title when title prop provided', () => {
    const matrix = ReactTestRenderer.create(
      <Matrix>
        <MatrixItem title="Title">
          description
        </MatrixItem>
      </Matrix>,
    );
    const json = matrix.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should convert title to link when href prop provided', () => {
    const matrix = ReactTestRenderer.create(
      <Matrix>
        <MatrixItem title="Title" href="#">
          description
        </MatrixItem>
      </Matrix>,
    );
    const json = matrix.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render an image when img prop provided', () => {
    const matrix = ReactTestRenderer.create(
      <Matrix>
        <MatrixItem title="Title" href="#" img={{ src: 'http://placehold.it/60x60', alt: 'icon' }}>
          description
        </MatrixItem>
      </Matrix>,
    );
    const json = matrix.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should auto-fill to have a multiple of three MatrixItem children', () => {
    const matrix = ReactTestRenderer.create(
      <Matrix>
        <MatrixItem>
          description
        </MatrixItem>
        <MatrixItem>
          description
        </MatrixItem>
        <MatrixItem>
          description
        </MatrixItem>
        <MatrixItem>
          description
        </MatrixItem>
      </Matrix>,
    );
    const json = matrix.toJSON();
    expect(json).toMatchSnapshot();
  });
});
