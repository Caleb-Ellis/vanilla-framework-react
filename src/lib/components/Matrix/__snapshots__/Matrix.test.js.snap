// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matrix component should auto-fill to have a multiple of three MatrixItem children 1`] = `
<ul
  className="p-matrix u-clearfix"
>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
</ul>
`;

exports[`Matrix component should convert title to link when href prop provided 1`] = `
<ul
  className="p-matrix u-clearfix"
>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      <h3
        className="p-matrix__title"
      >
        <a
          className="p-matrix__link"
          href="#"
        >
          Title
        </a>
      </h3>
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
</ul>
`;

exports[`Matrix component should render a Matrix with multiple MatrixItem correctly 1`] = `
<ul
  className="p-matrix u-clearfix"
>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
</ul>
`;

exports[`Matrix component should render a Matrix with one MatrixItem correctly 1`] = `
<ul
  className="p-matrix u-clearfix"
>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
</ul>
`;

exports[`Matrix component should render a title when title prop provided 1`] = `
<ul
  className="p-matrix u-clearfix"
>
  <li
    className="p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      <h3
        className="p-matrix__title"
      >
        Title
      </h3>
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
</ul>
`;

exports[`Matrix component should render an image when img prop provided 1`] = `
<ul
  className="p-matrix u-clearfix"
>
  <li
    className="p-matrix__item"
  >
    <img
      alt="icon"
      className="p-matrix__img"
      src="http://placehold.it/60x60"
    />
    <div
      className="p-matrix__content"
    >
      <h3
        className="p-matrix__title"
      >
        <a
          className="p-matrix__link"
          href="#"
        >
          Title
        </a>
      </h3>
      <div
        className="p-matrix__desc"
      >
        description
      </div>
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
  <li
    className="u-hide--small p-matrix__item"
  >
    <div
      className="p-matrix__content"
    >
      
      <div
        className="p-matrix__desc"
      />
    </div>
  </li>
</ul>
`;
